models/config_file_tree.go

```diff
@@ -0,0 +1,106 @@
+package models
+
+import (
+	"fmt"
+	"gopkg.in/yaml.v2"
+)
+
+func (ymlTree *ConfigFileTreeModel) Merge() (string, error) {
+	result, err := merge(ymlTree)
+	if err != nil {
+		return "", err
+	}
+
+	mergedYml, err := yaml.Marshal(result)
+	if err != nil {
+		return "", fmt.Errorf("failed to create YML result, error: %s", err)
+	}
+
+	return string(mergedYml), nil
+}
+
+type yamlMap = map[any]any
+
+func merge(ymlTree *ConfigFileTreeModel) (yamlMap, error) {
+	// Initial state is an empty map (YAML root)
+	initial := make(yamlMap)
+
+	result, err := mergeTree(initial, ymlTree)
+	if err != nil {
+		return nil, fmt.Errorf("failed to merge YML files, error: %s", err)
+	}
+
+	// Remove include list from result
+	delete(result, "include")
+
+	return result, nil
+}
+
+func mergeTree(existingValue yamlMap, treeToMerge *ConfigFileTreeModel) (yamlMap, error) {
+	var err error
+
+	// DFS: first the includes in the specified order, then the including file
+	for _, includedTree := range treeToMerge.Includes {
+		existingValue, err = mergeTree(existingValue, &includedTree)
+		if err != nil {
+			return nil, fmt.Errorf("failed to merge YML file %s, error: %s", includedTree.Path, err)
+		}
+	}
+
+	// We assume that each YML file has a map at root, it's invalid otherwise
+	var config yamlMap
+	err = yaml.Unmarshal([]byte(treeToMerge.Contents), &config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse YML file %s, error: %s", treeToMerge.Path, err)
+	}
+
+	if config == nil {
+		// File is empty
+		config = make(yamlMap)
+	}
+
+	return mergeMap(existingValue, config), nil
+}
+
+func mergeValue(existingValue any, valueToMerge any) any {
+
+	switch valueToMerge.(type) {
+	case yamlMap:
+		existingMap, ok := existingValue.(yamlMap)
```

Comment:
ok is always true, because we switch on the type of `valueToMerge` and in this case it is `yamlMap`, right?

Reply:
Not necessarily. It can (theoretically) happen that items have different types in YML modules, in that case, `existingValue` would be something else. One example where this is intended in our schema is the trigger map conditions which can be simple values (wildcard string) or maps (with a regex item) - these won't be merged now, but the point is, this can happen in YML and the merger should be able to handle it.

Reply:
Ahh OK, I mixed things here, now I see that we type switch on `valueToMerge` and type cast `existingValue`.